pkg/queryfrontend/queryinstant_codec.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package queryfrontend
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/thanos-io/thanos/internal/cortex/cortexpb"
+	"github.com/thanos-io/thanos/internal/cortex/querier/queryrange"
+	cortexutil "github.com/thanos-io/thanos/internal/cortex/util"
+	"github.com/thanos-io/thanos/internal/cortex/util/spanlogger"
+	queryv1 "github.com/thanos-io/thanos/pkg/api/query"
+)
+
+// queryInstantCodec is used to encode/decode Thanos instant query requests and responses.
+type queryInstantCodec struct {
+	partialResponse bool
+}
+
+// NewThanosQueryInstantCodec initializes a queryInstantCodec.
+func NewThanosQueryInstantCodec(partialResponse bool) *queryInstantCodec {
+	return &queryInstantCodec{
+		partialResponse: partialResponse,
+	}
+}
+
+func (c queryInstantCodec) MergeResponse(responses ...queryrange.Response) (queryrange.Response, error) {
+	if len(responses) == 0 {
+		return queryrange.NewEmptyPrometheusInstantQueryResponse(), nil
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	promResponses := make([]*queryrange.PrometheusInstantQueryResponse, 0, len(responses))
+	for _, resp := range responses {
+		promResponses = append(promResponses, resp.(*queryrange.PrometheusInstantQueryResponse))
+	}
+	res := &queryrange.PrometheusInstantQueryResponse{
+		Status: queryrange.StatusSuccess,
+		Data: queryrange.PrometheusInstantQueryData{
+			ResultType: model.ValVector.String(),
+			Result: queryrange.PrometheusInstantQueryResult{
+				Result: &queryrange.PrometheusInstantQueryResult_Samples{
+					Samples: vectorMerge(promResponses),
+				},
+			},
+			Stats: queryrange.StatsMerge(responses),
+		},
+	}
+	return res, nil
+}
+
+func (c queryInstantCodec) DecodeRequest(_ context.Context, r *http.Request, forwardHeaders []string) (queryrange.Request, error) {
+	var (
+		result ThanosQueryInstantRequest
+		err    error
+	)
+	if len(r.FormValue("time")) > 0 {
+		result.Time, err = cortexutil.ParseTime(r.FormValue("time"))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	result.Dedup, err = parseEnableDedupParam(r.FormValue(queryv1.DedupParam))
+	if err != nil {
+		return nil, err
+	}
+
+	if r.FormValue(queryv1.MaxSourceResolutionParam) == "auto" {
```

Comment:
Does downsampling apply to instant queries?

Reply:
Yeah from the instant query API I can see the param is used.